src/Microsoft.OData.Edm/Csdl/Parsing/SchemaJsonParser.cs

```diff
@@ -0,0 +1,1564 @@
+                        isOpen = propertyValue.ParseAsBoolean(context);
+                        break;
+
+                    case "$HasStream":
+                        // The value of $HasStream is one of the Boolean literals true or false. Absence of the member means false.
+                        hasStream = propertyValue.ParseAsBoolean(context);
+                        break;
+
+                    case "$Key":
+                        // The value of $Key is an array with one item per key property.
+                        csdlKey = ParseCsdlKey(propertyName, propertyValue, context);
+                        break;
+
+                    default:
+                        CsdlElement element = TryParseStructuredTypeElement(propertyName, propertyValue, context);
+                        if (element != null)
+                        {
+                            members.Add(element);
+                        }
+                        else
+                        {
+                            ReportUnexpectedElement(propertyValue, context);
+                        }
+
+                        break;
+                }
+            });
+
+
+            CsdlEntityType entityType = new CsdlEntityType(name,
+                baseType,
+                isAbstract == null ? false : isAbstract.Value, // default is false
+                isOpen == null ? false : isOpen.Value, // default is false
+                hasStream == null ? false : hasStream.Value, // default is false
+                csdlKey,
+                members.OfType<CsdlProperty>(),
+                members.OfType<CsdlNavigationProperty>(),
+                context.Location());
+
+            foreach (var annotation in members.OfType<CsdlAnnotation>())
+            {
+                entityType.AddAnnotation(annotation);
+            }
+
+            return entityType;
+        }
+
+        private static CsdlElement TryParseStructuredTypeElement(string name, JsonElement element, JsonParserContext context)
+        {
+            // It also MAY contain members representing structural properties and navigation properties as well as annotations.
+            Debug.Assert(!string.IsNullOrEmpty(name), "json reader should verify that the property name string is not null or empty.");
+            Debug.Assert(context != null);
+
+            if (AnnotationJsonParser.TryParseCsdlAnnotation(name, element, context, out CsdlAnnotation annotation))
+            {
+                return annotation;
+            }
+
+            // The remaining is property or navigation property, both them are required an object
+            if (element.ValueKind != JsonValueKind.Object)
+            {
+                return null;
+            }
+
+            // so it's complex property, Property is an object
+            string kind = GetKind(element, context);
+            if (kind == "NavigationProperty")

```

Comment:
>NavigationProperty [](start = 25, length = 18)

Should probably define constants for these string values

Reply:
I did it using a constant class. However, i changed my mind.

I think we use "string" directly in the code makes the code readable, and it's clearer than to use "ODataConstants.NavigationProperty". What's your thoughts?

---
In reply to: [443075887](https://github.com/OData/odata.net/pull/1775#discussion_r443075887) [](ancestors = 443075887)

Reply:
I'm okay with that, as long as we're consistent, and the string is not likely to change.

---
In reply to: [444538393](https://github.com/OData/odata.net/pull/1775#discussion_r444538393) [](ancestors = 444538393,443075887)